### PR TITLE
Update DNS to deal with edge nodes

### LINKED ISSUE #822: Simplify Terraform DNS configurations
The current DNS configuration modules reflect an older view of how the project needs to be connected up. Namely, we get records like this:

| type | IP | name |
|-------|-----|---------|
| A | `10.10.10.1` | `shortname-control-01.yourdomain.com` |
| A | `10.10.10.2` | `shortname-control-02.yourdomain.com` |
| A | `10.10.10.3` | `shortname-control-03.yourdomain.com` |
| A | `10.10.10.4` | `shortname-worker-001.yourdomain.com` |
| A | `10.10.10.4` | `*.shortname-lb.yourdomain.com` |

This made sense when we needed to connect to a specific control node for tasks, but since we have Mantl UI now, we can connect to any node. We also don't need to connect to worker nodes anymore as of #821, we now connect to edge nodes.

So I'd propose our DNS config should look like this:

| type | IP | name |
|------|----|------|
| A | `10.10.10.1` | `control.yourdomain.com` |
| A | `10.10.10.2` | `control.yourdomain.com` |
| A | `10.10.10.3` | `control.yourdomain.com` |
| A | `10.10.10.5` (edge-01) | `*.yourdomain.com` |

This should have a configurable subdomain, so if you want `control.mantl.yourdomain.com`, you could have that.

Thoughts?

### PR DESCRIPTION
**Pull request by BrianHicks**
Following the discussion on #822, here's the changes made to support the edge roles.

Closes #822

### CODE REVIEW

File: docs/getting_started/dns.rst [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+- ``mantl-control-03.node.example.com``
+- ``mantl-worker-001.node.example.com``
+- ``mantl-worker-002.node.example.com``
+- ``mantl-worker-003.node.example.com``
+- ``mantl-worker-004.node.example.com``
+- ``mantl-edge-01.node.example.com``
+- ``mantl-edge-02.node.example.com``
+- ``control.example.com`` (pointing to control 1)
+- ``control.example.com`` (pointing to control 2)
+- ``control.example.com`` (pointing to control 2)
```
Comment (ChrisAubuchon) [inline docs/getting_started/dns.rst]: Is this supposed to be `(pointing to control 3)`?

Comment (BrianHicks) [inline docs/getting_started/dns.rst]: yep! Thanks.

File: terraform/route53/dns/main.tf [hunk 2]
```diff
@@ -18,16 +33,27 @@ resource "aws_route53_record" "dns-control" {
...
+  type = "A"
+  ttl = 60
+}
+
+# group records
+resource "aws_route53_record" "dns-control-group" {
+  count = "${var.control_count}"
+  zone_id = "${var.hosted_zone_id}"
+  name = "${var.control_subdomain}${var.subdomain}.${var.domain}"
+  records = ["${elements(split(",", var.control_ips), count.index)}"]
```
Comment (ryane) [inline terraform/route53/dns/main.tf]: shouldn't this be `element`?

File: terraform/route53/dns/main.tf [hunk 3]
```diff
@@ -18,16 +33,27 @@ resource "aws_route53_record" "dns-control" {
...
 }
 
-resource "aws_route53_record" "dns-worker-haproxy" {
+resource "aws_route53_record" "dns-wildcard" {
+  count = "${var.edge_count}"
   zone_id = "${var.hosted_zone_id}"
-  name = "*.${var.short_name}-lb.${var.domain}"
-  records = ["${split(\",\", var.worker_ips)}"]
+  name = "*${var.subdomain}.${var.domain}"
+  records = ["${elements(split(",", var.edge_ips), count.index)}"]
```
Comment (ryane) [inline terraform/route53/dns/main.tf]: same here

Comment (ryane) [inline terraform/route53/dns/main.tf]: actually, this shouldn't use `element` as it takes a list. right now, you end up with a single edge node in the wildcard record

**Comment by ryane** [comment]
this is working on DNSimple but I noticed a couple of problems with the Route 53 setup.

File: terraform/gce/dns/main.tf [hunk 4]
```diff
@@ -1,33 +1,59 @@
...
-resource "google_dns_record_set" "dns-worker-haproxy" {
+resource "google_dns_record_set" "dns-wildcard" {
+  count = "${var.edge_count}"
   managed_zone = "${var.managed_zone}"
-  name = "${var.short_name}-lb.${var.domain}."
+  name = "*${var.subdomain}.${var.domain}."
   type = "A"
   ttl = 60
-  rrdatas = ["${split(\",\", var.worker_ips)}"]
+  rrdatas = ["${element(split(",", var.edge_ips), count.index)}"]
```
Comment (ryane) [inline terraform/gce/dns/main.tf]: I am not setup to test google cloud dns but it looks you might not want to use `element` here either.

File: terraform/route53/dns/main.tf (hunk 3 above)
Comment (BrianHicks) [inline terraform/route53/dns/main.tf]: yeah... I'm not set up to test this provider, so I just tried not to make mistakes and hoped someone would catch them in review. Fortunately, you did. Thanks!

**Comment by BrianHicks** [comment]
@ryane fixed those problems, did you see anything else?

**Comment by ryane** [comment]
looks good. confirmed it is working on route53. gce *looks* ok afaict

**Pull request merged**